sale_invoice_plan/models/sale.py

```diff
@@ -0,0 +1,223 @@
+# Copyright 2019 Ecosoft Co., Ltd (http://ecosoft.co.th/)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html)
+from dateutil.relativedelta import relativedelta
+from odoo import models, fields, api, _
+from odoo.exceptions import UserError, ValidationError
+from odoo.addons import decimal_precision as dp
+from odoo.tools.float_utils import float_round as round
+
+
+class SaleOder(models.Model):
+    _inherit = 'sale.order'
+
+    invoice_plan_ids = fields.One2many(
+        comodel_name='sale.invoice.plan',
+        inverse_name='sale_id',
+        string='Inovice Plan',
+        copy=False,
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+    use_invoice_plan = fields.Boolean(
+        string='Use Invoice Plan',
+        default=False,
+        copy=False,
+        readonly=True,
+        states={'draft': [('readonly', False)]},
+    )
+    ip_invoice_plan = fields.Boolean(
+        string='Invoice Plan In Process',
+        compute='_compute_ip_invoice_plan',
+        help="At least one invoice plan line pending to create invoice",
+    )
+
+    @api.multi
+    def _compute_ip_invoice_plan(self):
+        for rec in self:
+            rec.ip_invoice_plan = rec.use_invoice_plan and \
+                rec.invoice_plan_ids and \
+                len(rec.invoice_plan_ids.filtered(lambda l: not l.invoiced))
+
+    @api.constrains('state')
+    def _check_invoice_plan(self):
+        for rec in self:
+            if rec.state != 'draft':
+                if rec.invoice_plan_ids.filtered(lambda l: not l.percent):
+                    raise UserError(
+                        _('Please fill percentage for all invoice plan lines'))
+
+    @api.multi
+    def action_confirm(self):
+        for rec in self:
```

Comment:
You could use this :

```suggestion
        if self.filtered(lambda r: r.use_invoice_plan and not r.invoice_plan_ids):
            raise...
```

just to avoid indent

Reply:
Done